### DISCUSSION TITLE
What's the recommended approach to extend the `account-baseline` Terraform modules?

### DISCUSSION THREAD

**Discussion by p5** [gruntwork-io/knowledge-base#640] [How do I]
There are various changes I wish to make on all accounts in the reference architecture - including:
- Enabling AWS Inspector
- Enabling AWS CIS v1.4.0 in Security Hub
- Enabling PCI DSS in Security Hub

Should we wrap the `account-baseline-app` modules in our own custom module where we can implement these, or is there a different recommended approach (like creating completely separate modules)?

And if wrapping the modules is recommended, what's the process for switching to these?  Will we need to run `terraform import` for each resource - which will be hundreds of resources in `account-baseline`?  

---

<ins datetime="2023-01-19T12:27:04Z">
  <p><a href="https://support.gruntwork.io/hc/requests/109811">Tracked in ticket #109811</a></p>
</ins>

**Comment by josh-padnick** [comment]
When you're working with third-party modules like Gruntwork or open source, you always have three options for consuming them:

1. Use the module directly
2. Wrap the module
3. Fork the module

Each approach has its pros/cons, but generally you're trading control for convenience. There is of course a 4th option, which is to create a new module entirely.

It's best practice for Terraform modules to be small and modular, but the Landing Zone module is our one module that most challenges that pattern by stuffing, as you pointed out yourself, quite a lot of resources into one module. Working with large modules means long `terraform plan` and `terraform apply` times and introduces a wider surface area that could cause bugs.

Given that, for your needs, I would recommend creating a new module called something like `account-baseline-app-acme` which you add to all new accounts in addition to the Gruntwork Landing Zone module. Admittedly, it makes me sad to have two modules to configure one account, but this gives you a clean way to "append" functionality to your new accounts without the baggage of a large module. The one thing you need to be careful of is not to duplicate functionality in the Gruntwork Landing Zone module.

I'd like to get one additional opinion here, so I'll check with a colleague as well on this.
